src/audio/chain_dma.c

```diff
@@ -0,0 +1,592 @@
+	struct chain_dma_data *cd;
+	struct comp_dev *dev;
+	int ret;
+
+	if (host_dma_id >= max_chain_number)
+		return NULL;
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		goto error1;
+
+	cd->first_data_received = false;
+	cd->under_over_run_notification_sent = false;
+	cd->cs = scs ? 2 : 4;
+	cd->chain_task.state = SOF_TASK_STATE_INIT;
+
+	comp_set_drvdata(dev, cd);
+
+	ret = chain_task_init(dev, host_dma_id, link_dma_id, fifo_size);
+	if (ret)
+		goto error2;
+
+	return dev;
+
+error2:
+	rfree(cd);
+error1:
+	rfree(dev);
+	return NULL;
+}
+
+int chain_task_start(struct comp_dev *dev)
+{
+	struct comp_driver_list *drivers = comp_drivers_get();
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	k_spinlock_key_t key;
+	int ret;
+
+	comp_info(dev, "chain_task_start(), host_dma_id = 0x%08x", cd->host_connector_node_id.dw);
+
+	key = k_spin_lock(&drivers->lock);
+	switch (cd->chain_task.state) {
+	case SOF_TASK_STATE_QUEUED:
+		k_spin_unlock(&drivers->lock, key);
+		return 0;
+	case SOF_TASK_STATE_COMPLETED:
+		break;
+	case SOF_TASK_STATE_INIT:
+		break;
+	default:
+		comp_err(dev, "chain_task_start(), bad state transition");
+		goto error1;
+	}
+	cd->under_over_run_notification_sent = false;
+
+	if (cd->stream_direction == SOF_IPC_STREAM_PLAYBACK) {
+		ret = chain_host_start(dev);
+		if (ret)
+			goto error1;
+		ret = chain_link_start(dev);
+		if (ret) {
+			chain_host_stop(dev);
+			goto error1;
+		}
+	} else
+	{
+		ret = chain_link_start(dev);
+		if (ret)
+			goto error1;
+		ret = chain_host_start(dev);
+		if (ret) {
+			chain_link_stop(dev);
+			goto error1;
+		}
+	}
+
+	pm_policy_state_lock_get(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
+
+	ret = schedule_task(&cd->chain_task, 0, 0);
+	if (ret)
+		goto error1;
+
+	cd->chain_task.state = SOF_TASK_STATE_INIT;
+	k_spin_unlock(&drivers->lock, key);
+
+	return 0;
+
+error1:
+	k_spin_unlock(&drivers->lock, key);
+	return -EINVAL;
+}
+
+int chain_task_pause(struct comp_dev *dev)
+{
+	struct comp_driver_list *drivers = comp_drivers_get();
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	k_spinlock_key_t key;
+	int ret;
+
+	if (cd->chain_task.state == SOF_TASK_STATE_FREE)
+		return 0;
+
+	key = k_spin_lock(&drivers->lock);
+	cd->first_data_received = false;
+	if (cd->stream_direction == SOF_IPC_STREAM_PLAYBACK)
+	{
+		ret = chain_host_stop(dev);
+		if (ret < 0)
+			goto error1;
+		ret = chain_link_stop(dev);
+		if (ret < 0)
+			goto error2;
+	} else {
+		ret = chain_link_stop(dev);
+		if (ret < 0)
+			goto error1;
+		ret = chain_host_stop(dev);
+		if (ret < 0)
+			goto error3;
+	}
+
+	cd->chain_task.state = SOF_TASK_STATE_COMPLETED;
+	schedule_task_free(&cd->chain_task);
+	pm_policy_state_lock_put(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
+	k_spin_unlock(&drivers->lock, key);
+
+	return 0;
+error2:
+	chain_host_stop(dev);
+	goto error1;
+error3:
+	chain_link_stop(dev);
+	goto error1;
```

Comment:
This does not make sense:
you jump here from line 155 _after_ the `chain_link_stop()` has been successful but the `chain_host_stop()` failed and you do another `chain_link_stop()`?